audio-worklet/src/noisegate-script-processor.js

```diff
@@ -0,0 +1,187 @@
+   * is above a specified threshold.
+   * @param {BaseAudioContext} context the audio context
+   * @param {Number} options.channels the number of input and output
+   *                                  channels, a maximum of two
+   * @param {Object} options parameters for the noise gate
+   * @param {Number} options.attack seconds for gate to fully close
+   * @param {Number} options.release seconds for gate to fully open
+   * @param {Number} options.bufferSize the size of an onaudioprocess window
+   * @param {Number} options.alpha seconds for envelope follower's smoothing
+   *                                   filter alpha weight
+   * @param {Number} options.threshold decibel level beneath which sound is
+   *                                   muted
+   */
+  constructor(context, options) {
+    if (!(context instanceof BaseAudioContext))
+      throw 'Not a valid audio context.';
+    if (!options) options = {};
+    const numberOfChannels = options.numberOfChannels || 1;
+    if (numberOfChannels > 2)
+      throw 'The maximum supported number of channels is two.';
+
+    this.context_ = context;
+    const bufferSize = options.bufferSize || 0;
+    this.threshold = options.threshold || 0;
+    this.attack = options.attack || 0;
+    this.release = options.release || 0;
+
+    // Alpha controls a tradeoff between the smoothness of the
+    // envelope and its delay, with a higher value giving more smoothness at
+    // the expense of delay and vice versa. The time constant of the filter
+    // has been set experimentally to minimize delay while still adequately
+    // suppressing high frequency oscillation.
+    this.alpha_ = this.getTimeConstant_(0.0025);
+   
+    this.node_ = this.context_.createScriptProcessor(
+        bufferSize, numberOfChannels, numberOfChannels);
+    this.node_.onaudioprocess = this.onaudioprocess_.bind(this);
+
+    // The noise gate is connected to and from by dummy input and output nodes.
+    this.input = new GainNode(this.context_);
+    this.output = new GainNode(this.context_);
+    this.input.connect(this.node_).connect(this.output);
+
+    // The last envelope level of a given buffer.
+    this.lastLevel_ = 0;
+
+    // The last weight (between 0 and 1) assigned, where 1 means the gate
+    // is open and 0 means it is closed and the sample in the output buffer is
+    // muted.
+    this.lastWeight_ = 1.0;
+    this.channel_ = new Float32Array(this.node_.bufferSize);
+    this.envelope_ = new Float32Array(this.node_.bufferSize);
+    this.weights_ = new Float32Array(this.node_.bufferSize);
+  }
+
+  /**
+   * Gradually mutes input which registers beneath specified threshold.
+   * @param {AudioProcessingEvent} event event object containing
+   *                                     input and output buffers
+   */
+  onaudioprocess_(event) {
+    // Stereo input is downmixed to mono input via averaging.
+    if (event.inputBuffer.numberOfChannels === 2) {
```

Comment:
Beyond what is already there? '// Stereo input is downmixed to mono input via averaging.'